src/xwrap.rs

```diff
@@ -243,19 +279,89 @@ impl<'a> Drop for ScreenRectIter<'a> {
     }
 }
 
-pub fn parse_geometry(g: ffi::CString) -> util::Rect {
-    unsafe {
-        let mut x = 0;
-        let mut y = 0;
-        let mut w = 0;
-        let mut h = 0;
-        xlib::XParseGeometry(g.as_ptr() as *const raw::c_char, &mut x, &mut y, &mut w, &mut h);
-
-        util::Rect {
+/// Parse the Area to capture.
+///
+/// The format is `WxH+X+Y` to fully specify the format of `man 3 XParseGeometry`
+/// `[=][<width>{xX}<height>][{+-}<xoffset>{+-}<yoffset>]`
+pub fn parse_geometry(g: &str) -> Result<util::Rect, ParseError> {
+    let g = CString::new(g).expect("Failed to convert CString");
+    let mut x = i32::MIN;
+    let mut y = i32::MIN;
+    let mut w = u32::MIN;
+    let mut h = u32::MIN;
+
+    let ret: i32 = unsafe {
+        xlib::XParseGeometry(
+            g.as_ptr() as *const raw::c_char,
+            &mut x,
+            &mut y,
+            &mut w,
+            &mut h,
+        )
+    };
+    const ALL_VALUES: i32 = xlib::YValue | xlib::XValue | xlib::WidthValue | xlib::HeightValue;
+    const WH_VALUES: i32 = xlib::WidthValue | xlib::HeightValue;
+
+    match ret {
+        ALL_VALUES => Ok(util::Rect {
             x: x,
             y: y,
             w: w as i32,
             h: h as i32,
-        }
+        }),
+        xlib::NoValue => Err(ParseError::NoValue),
+        WH_VALUES => Err(ParseError::MissingXY),
```

Comment:
I think we should preserve the original error, maybe keep `X` and `Y` separate since that is what was returned by the C function?